File: pypp_cli/test_dir/cpp/glfw_test.cpp
#include "cstdlib"
#include "exceptions/exception.h"
#include "py_str.h"
#include "pypp_util/main_error_handler.h"
#include <GLFW/glfw3.h>

void glfw_test() {
    if (!glfwInit()) {
        throw pypp::Exception(pypp::PyStr("Failed to initialize GLFW"));
    }
    GLFWwindow *window = glfwCreateWindow(
        640, 480, pypp::PyStr("Hello World").str().c_str(), NULL, NULL);
    if (!window) {
        glfwTerminate();
        throw pypp::Exception(pypp::PyStr("Failed to create GLFW window"));
    }
    glfwMakeContextCurrent(window);
    while (!glfwWindowShouldClose(window)) {
        glfwSwapBuffers(window);
        glfwPollEvents();
    }
    glfwTerminate();
}

int main() {
    try {
        glfw_test();
        return 0;
    } catch (...) {
        pypp::handle_fatal_exception();
        return EXIT_FAILURE;
    }
}

File: pypp_cli/test_dir/python/glfw_test.py
import glfw
from pypp_bridge_lib_glfw import GLFWwindowPtr
from pypp_python import to_c_string, NULL


def glfw_test():
    # Initialize the library
    if not glfw.init():
        raise Exception("Failed to initialize GLFW")
    # Create a windowed mode window and its OpenGL context
    window: GLFWwindowPtr = glfw.create_window(
        640, 480, to_c_string("Hello World"), NULL, NULL
    )
    if not window:
        glfw.terminate()
        raise Exception("Failed to create GLFW window")

    # Make the window's context current
    glfw.make_context_current(window)

    # Loop until the user closes the window
    while not glfw.window_should_close(window):
        # Render here, e.g. using pyOpenGL

        # Swap front and back buffers
        glfw.swap_buffers(window)

        # Poll for and process events
        glfw.poll_events()

    glfw.terminate()


if __name__ == "__main__":
    glfw_test()
